File: frontend/src/schemas/problem.ts
import { z } from "zod";

import { dependencyStatusesSchema } from "./health";

export const problemDetailsSchema = z
  .object({
    type: z.string(),
    title: z.string(),
    status: z.number().int().min(400).max(599),
    detail: z.string(),
    instance: z.string(),
    trace_id: z.string().regex(/^[0-9a-f]{32}$/),
    dependencies: dependencyStatusesSchema.nullable().optional(),
  })
  .strict();

export type ProblemDetails = z.infer<typeof problemDetailsSchema>;

File: src/invoiceops_agent/api/schemas/problem.py
"""RFC 7807 problem details with correlation and optional readiness extensions."""

from pydantic import BaseModel, ConfigDict, Field

from invoiceops_agent.api.schemas.health import DependencyStatuses


class ProblemDetails(BaseModel):
    model_config = ConfigDict(extra="forbid")

    type: str = "about:blank"
    title: str
    status: int = Field(ge=400, le=599)
    detail: str
    instance: str
    trace_id: str = Field(pattern=r"^[0-9a-f]{32}$")
    dependencies: DependencyStatuses | None = None
